lighthouse-core/config/index.js

```diff
@@ -0,0 +1,161 @@
+    const filteredAudits = audits.filter(a => {
+      // If there is no whitelist, assume all.
+      if (!whitelist) {
+        return true;
+      }
+
+      const auditName = a.toLowerCase();
+      const inWhitelist = whitelist.has(auditName);
+
+      if (!inWhitelist) {
+        rejected.push(auditName);
+      }
+
+      return inWhitelist;
+    });
+
+    if (rejected.length) {
+      log.log('info', 'Running these audits:', `${filteredAudits.join(', ')}`);
+      log.log('info', 'Skipping these audits:', `${rejected.join(', ')}`);
+    }
+
+    return filteredAudits;
+  }
+
+  static expandAudits(audits) {
+    // It's possible we didn't get given any audits (but existing audit results), in which case
+    // there is no need to do any expansion work.
+    if (!audits) {
+      return;
+    }
+
+    return audits.map(audit => {
+      // If this is already instantiated, don't do anything else.
+      if (typeof audit !== 'string') {
+        return audit;
+      }
+
+      try {
+        return require(`../audits/${audit}`);
+      } catch (requireError) {
+        throw new Error(`Unable to locate audit: ${audit}`);
+      }
+    });
+  }
+
+  static getGatherersNeededByAudits(config) {
+    // It's possible we didn't get given any audits (but existing audit results), in which case
+    // there is no need to do any work here.
+    if (!config.audits) {
+      return new Set();
+    }
+
+    return config.audits.reduce((list, audit) => {
+      audit.meta.requiredArtifacts.forEach(artifact => list.add(artifact));
+      return list;
+    }, new Set());
+  }
+
+  static filterPasses(config) {
+    const requiredGatherers = ConfigParser.getGatherersNeededByAudits(config);
+
+    // Make sure we only have the gatherers that are needed by the audits
+    // that have been listed in the config.
+    const filteredPasses = config.passes.map(pass => {
+      pass.gatherers = pass.gatherers.filter(gatherer => {
+        if (typeof gatherer !== 'string') {
+          return requiredGatherers.has(gatherer.name);
+        }
+
+        try {
+          const GathererClass = Driver.getGathererClass(gatherer);
+          return requiredGatherers.has(GathererClass.name);
+        } catch (requireError) {
+          throw new Error(`Unable to locate gatherer: ${gatherer}`);
+        }
+      });
+
+      return pass;
+    })
+
+    // Now remove any passes which no longer have gatherers.
+    .filter(p => p.gatherers.length > 0);
+    return filteredPasses;
+  }
+
+  static parsePerformanceLog(logs) {
+    // Parse logs for network events
+    const networkRecords = recordsFromLogs(logs);
+
+    // Use critical request chains gatherer to create the critical request chains artifact
+    const criticalRequestChainsGatherer = new CriticalRequestChainsGatherer();
+    criticalRequestChainsGatherer.afterPass({}, {networkRecords});
+
+    return criticalRequestChainsGatherer.artifact;
+  }
+
+  static parse(config, auditWhitelist) {
```

Comment:
ideally we want this to return a different object than the `config` that comes in. The biggest reason would be that when running lighthouse as a module, you don't expect the simple config object to come back with properties replaced with live objects usable only internally to lighthouse. It would also remove the need for some of the work arounds in this file, like checking if an audit is a string or has already been instantiated, for instance.

The easiest way to do this might just be a deep copy of the passed in config object, and replace audit names with instantiated objects on the copy. This should be relatively easy (in theory) because the config should have just been parsed from JSON and so the only parts we'll care about have all the simplicity required by JSON.

However, it might also require baking some knowledge of the structure of the config in the deep copy function, which means lots of changing code in multiple places in the future as things are added to or removed from the config.

Another alternative would be to treat the config we have now as a read-only config object and any modifications that would have been made to it would actually go into a second object, a core/runtime/pipeline whatever object that houses all the actual gatherer instances that will be used, the audit instances, etc. That way we preserve the config object passed in, don't have to muck around with a deep copy, and new code could still add or remove options to the config file without unrelated code having to be altered to handle the changes.